`DoFHandler`: Support virtual degrees of freedom

tamiko opened the pull request:
@kronbichler @luca-heltai @peterrum This is what I have in mind.

You can find a modified `step-40` in the build directory that I used to test whether this works - and yes, so far no problems.

The changes required are surprisingly minimal. In essence we only need to:
 - shift the global enumeration of the local range
 - make the virtual dof range available in some form

I looked into some alternatives (including playing this index game in the `Partitioner`). But I think adding support for this in the `DoFHandler` is the most elegant solution.

The interface here is a quick 30 minute hack testing the core functionality.
We should probably design something much more elegant. (including an interface to create multigrid structures.)

But I wanted to share this here early on. Because it works surprisingly well.

pcafrica committed:
Merge SolutionTransfer and p::d::SolutionTransfer

tamiko committed:
DoFHandler: add interface for virtual degrees of freedom

tamiko committed:
TEMPORARY

kronbichler commented:
I think the code is indeed very nice. One of the pros is that it is actually the `DoFHandler`'s task to organize degrees of freedom, irrespective of where they come from.

One of the topics I already mentioned in our discussion board is the question about ghosting. This is of course up to the user code to decide whether and to which extent the virtual DoFs should be accessible outside of the owner process. So it is not the business of the `DoFHandler` but the partitioner and parallel vectors, in order to ensure efficient communication routines.

luca-heltai commented:
Sorry, I'm late to the party. I like this option, but I think we still need to add support for non-local dofs. In this particular example, I see that you are essentially replicating the "dofs_offset" concept we had some time ago (done in a much better way, I must admit!) and that was removed since it was not very well tested. 
This won't work with basis functions associated to these dofs on some cells, in my understanding, right? You have to handle them manually all-together.
I think we still need this, and I'm in favour of this PR.

luca-heltai commented on include/deal.II/dofs/dof_handler.h:
```diff
@@ -713,13 +713,15 @@ class DoFHandler : public Subscriptor
   void
-  distribute_dofs(const FiniteElement<dim, spacedim> &fe);
+  distribute_dofs(const FiniteElement<dim, spacedim>    &fe,
+                  const dealii::types::global_dof_index &virtual_dofs = 0);
```
Do I understand correctly that these should be interpreted as the *total* number of virtual dofs? In your example, you are making all of them locally owned by all processors. This does not sound right to me, i.e., in the example above, if you run on *n* processors you should end up with a matrix that has `n` on the diagonal on the virtual dofs block.
Maybe we should add a constructor that actually takes an `IndexSet` as virtual dofs.

tamiko replied:
@luca-heltai Yes right now I only explored what this parameter should mean for serial and parallel::distributed (and not for the others). We'll have to discuss how the interface should look like anyway.

I really don't like this additional argument for `distribute_dofs` and I think it should either be a part of the finite element or a different method/interface.

AE7TB99 committed:
Update step-49, step-50 and their diffs.
Closes #17249.

masterleinad committed:
Remove CUDAWrappers, related documentation and macros

marcfehling committed:
Add remaining examples to tests as run_only.

gassmoeller committed:
Typo configuration

bangerth committed:
Add more C++ resources to step-1.

marcfehling committed:
Check that diff file exists for every example.

tjhei committed:
enable c++ standard library debug checks

ChrisPec27 committed:
removed last four void casts and replaced with maybe_unused

mkghadban committed:
replace (void) casts with [[maybe_unused]]

jerett-cc committed:
fix void pointer issue

landinjm committed:
marked potentially unused with [[maybe_unused]]

OresteMarquis committed:
deal.ii workshop tutorials on PR, tria (5-10) [[maybe_unused]] fix

eterrell11 committed:
Removed void casts with maybe unused

tjhei committed:
portable::FEEvaluation documentation fixes

magmaxt committed:
Replaced (void) cast with maybe unused | Modernize void cast

Commit:
Removed maybe unused in [[ ]]

tamiko committed:
run indent

ChrisPec27 committed:
Fixed error in attribute placement

tjhei committed:
MGTransferMF: fix nullptr access

transfer[min_level] is never used and okay to be null

tamiko committed:
indent

tamiko committed:
indent

tjhei committed:
fix test: invalid memory access

peterrum committed:
Support multiple components in Portable::MatrixFree.

AE7TB99 committed:
Update tests/examples/step-49.diff

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

jerett-cc committed:
indent and add [[maybe_unused]] to parameter list

gassmoeller committed:
Exclude test output from check

marcfehling committed:
Added default parameter file for step-71.

tjhei committed:
step-37: use MatrixFreeTools::compute_diagonal()

masterleinad committed:
Fix undefined behavior when advancing nullptr

tamiko committed:
Adolc/Symengine: always instantiate complex valued variants

tjhei committed:
Merge pull request #17489 from mkghadban/fix_something_funcion_lib

Use `[[maybe_unused]]` instead of `(void)` cast

tjhei committed:
Merge pull request #17490 from landinjm/maybe_unused_cpp17

Removed first four void casts in `source/distributed/tria.cc` and replaced them with `[[maybe_unused]]`

tjhei committed:
cleanup maybe-unused in fe.cc

followup to #17508

tjhei committed:
Merge pull request #17502 from magmaxt/remove_void_cast_grid_tools_cc

Use `[[maybe_unused]]` instead of void cast

masterleinad committed:
Merge pull request #17492 from tjhei/portable-mf-evaluation-flags

portable::FEEvaluation documentation fixes

tjhei committed:
Merge pull request #17495 from eterrell11/remove_void_cast_manifold

Replaced void casts with maybe unused

AE7TB99 committed:
Revert "Update tests/examples/step-49.diff"

This reverts commit 96801c77d0b6faa79b63aa701cfc24747d9860f1.

drwells committed:
Don't use std::uncaught_exception().

This is a C++17 feature which was shipped with GCC-6: we can get rid of the
workaround now.

drwells committed:
std::uncaught_exceptions(): use > 0, not != 0.

This is more consistent.

drwells committed:
Merge pull request #17485 from tjhei/disable-std-lib-checks

enable c++ standard library debug checks

tamiko committed:
Merge pull request #17486 from marcfehling/test-examples

Test all examples.

tjhei committed:
Merge pull request #17509 from jerett-cc/remove_void_cast_function_lib

Remove void cast function lib replace with [[maybe_unused]] attribute

tamiko committed:
Test symengine/batch_optimizer_01*: fix linkage by including vector.templates.h

drwells committed:
DataOutInterface: remove some reinterpret_casts.

drwells committed:
TensorProductPolynomials: remove some reinterpret_casts.

drwells committed:
TensorProductPolynomialsBubbles: remove some reinterpret_casts.

tamiko committed:
Merge pull request #17482 from gassmoeller/add_typo_check

Add typo check to CI

drwells committed:
ScalarPolynomialsBase: remove some reinterpret_casts.

drwells committed:
PolynomialSpace: remove some reinterpret_casts.

drwells committed:
PolynomialsRannacherTurek: remove some reinterpret_casts.

drwells committed:
Change a variable name in Assert().

This should fix the typo check failure.

farriscr18 committed:
Changed void cast to [[maybe_unused]]

tjhei committed:
Merge pull request #17519 from drwells/asser-to-assert

Change a variable name in Assert().

masterleinad committed:
Merge pull request #17515 from drwells/uncaught-exceptions

Consistently use std::uncaught_exceptions().

masterleinad committed:
Merge pull request #17512 from masterleinad/fix_ub_nullptt

Fix undefined behavior when advancing nullptr

masterleinad committed:
Merge pull request #17514 from tjhei/less-maybe-unused

cleanup maybe-unused in fe.cc

masterleinad committed:
Merge pull request #17461 from AE7TB99/update-step-49-step-50

Update step-49, step-50 and their diffs.

tamiko committed:
DoFHandler: renumber virtual dofs

tamiko committed:
TEMPORARY

tamiko committed:
Merge pull request #17513 from tamiko/fix_symengine_tests

Adolc/Symengine: always instantiate complex valued variants

kronbichler committed:
Merge pull request #17517 from tjhei/tutorials-compute-diagonal

step-37: use MatrixFreeTools::compute_diagonal()

tamiko committed:
Assert: make release mode variant work in constexpr context

bangerth committed:
Add a note to step-86.

bangerth committed:
Improve an error message.

masterleinad committed:
Fix test

tamiko committed:
Assert: guard new release variant with DEAL_II_HAVE_CXX20

tamiko committed:
CMake: set -Wno-unused-parameter in release mode

masterleinad committed:
Portable MatrixFree: use if constexpr for hanging nodes implementation

tjhei committed:
Merge pull request #17529 from bangerth/error

Improve an error message.

tjhei committed:
Merge pull request #17518 from drwells/remove-reinterpret-casts

Remove a lot of reinterpret_casts.

drwells committed:
Merge pull request #17483 from bangerth/1

Add more C++ resources to step-1.

marcfehling committed:
Fix a typo test.

peterrum committed:
Merge pull request #15706 from pcafrica/serialization_solution_transfer

Merge `SolutionTransfer` and `p::d::SolutionTransfer`

bangerth committed:
Merge pull request #17528 from bangerth/86

Add a note to step-86.

bangerth committed:
Merge pull request #17524 from tamiko/fix_release_mode

Assert: release mode fixes: constexpr context, disable warnings in release mode

gassmoeller committed:
Merge pull request #17487 from ChrisPec27/remove_void_cast_distributed_tria

Use `[[maybe_unused]]` instead of `(void)` cast

gassmoeller committed:
Merge pull request #17497 from OresteMarquis/remove_void_cast_tria_5to10

Remove void casts and use `[[maybe_unused]]` instead

drwells committed:
TensorProductPolynomials: remove some reinterpret_casts.

drwells committed:
AffineConstraints: remove some reinterpret_casts.

tjhei committed:
Merge pull request #17531 from marcfehling/fix_prm26

Fix a typo test.

gassmoeller committed:
Update some uses of maybe unused

farriscr18 committed:
Addressed comments

marcfehling committed:
Merge pull request #17530 from masterleinad/use_if_constexpr_matrix_free_kokkos

Portable MatrixFree: use if constexpr for hanging nodes implementation

tjhei committed:
fix compilation warning (unused arg)

masterleinad committed:
Fix compiling SolutionTransfer with complex values

tjhei committed:
Merge pull request #17534 from drwells/use-more-if-constexpr

Get rid of some more reinterpret_casts.

tjhei committed:
Merge pull request #17538 from masterleinad/fix_solution_transfer_complex

Fix compiling SolutionTransfer with complex values

tjhei committed:
one more

tamiko committed:
Merge pull request #17537 from tjhei/fix-compile-warn

fix compilation warning (unused arg)

drwells committed:
Drop support for p4est versions older than 2.2.

We already have several features which depend on this version.

marcfehling committed:
Merge pull request #17539 from drwells/p4est-2.2

Drop support for p4est versions older than 2.2.

drwells committed:
Improve our 32-bit vs 64-bit PETSc documentation.

Sanjeeb-Poudel committed:
Replaced void cast with maybe_unused

npcantrell committed:
removed (void) occurrences from source/fe/fe_face.cc

masterleinad committed:
Merge pull request #17525 from peterrum/kokkos_multi_componenent

Support multiple components in `Portable::MatrixFree`

tjhei committed:
Merge pull request #17504 from Sanjeeb-Poudel/remove_void_tria_accessor.cc

Replaced void cast with maybe_unused

kronbichler committed:
MPI allreduce: Do not invoke MPI function for MPI_COMM_SELF

kronbichler committed:
Fix typo: residuum -> residual

masterleinad committed:
Merge pull request #17546 from kronbichler/fix_typo2

Fix typo: residuum -> residual

tjhei committed:
Merge pull request #17545 from kronbichler/simplify_mpi_comm_self

MPI sum: Do not invoke MPI function for MPI_COMM_SELF

tjhei committed:
Merge pull request #17503 from PapaKwansa/remove_void_cast_mpi

Remove void cast mpi

gassmoeller committed:
Merge pull request #17501 from farriscr18/remove_void_cast_function_lib

Use `[[maybe_unused]]` instead of void cast

tjhei committed:
Merge pull request #17541 from drwells/document-petsc-int

Improve our 32-bit vs 64-bit PETSc documentation.

tjhei committed:
Merge pull request #17543 from npcantrell/removed_void_cast_in_source/fe/fe_face.cc

removed (void) occurrences from source/fe/fe_face.cc

marcfehling committed:
Fix some unused warnings.

masterleinad committed:
Always use Kokkos::abort when it's noreturn

Rombur committed:
Merge pull request #17474 from masterleinad/remove_cudawrappers

Remove CUDAWrappers, related documentation and macros

tamiko committed:
INCOMPLETE REFACTORING

gassmoeller committed:
Merge pull request #17548 from marcfehling/unused

Fix some unused warnings.

tjhei committed:
Merge pull request #17532 from gassmoeller/update_maybe_unused

Update some uses of [[maybe unused]]

peterrum committed:
Check for Fortran again

marcfehling committed:
Merge pull request #17552 from masterleinad/kokkos_noreturn_abort

Always use Kokkos::abort when it's noreturn

tamiko committed:
Merge pull request #17560 from peterrum/fortran

Check for Fortran again
